File: app/src/main/java/com/compuspulse/data/remote/model/Article.java
package com.compuspulse.data.remote.model;

import com.compuspulse.utils.HtmlUtils;
import com.google.gson.annotations.SerializedName;

import java.io.Serializable;

public class Article implements Serializable {
    private int id;
    private String title;
    private String summary;
    private String source;
    private String sourceUrl;
    private String publishTime;
    private String imageUrl;
    private String category;
    private int commentCount;
    private boolean favorite;

    @SerializedName("desc")
    private String apiDesc;
    @SerializedName("author")
    private String author;
    @SerializedName("shareUser")
    private String shareUser;
    @SerializedName("link")
    private String link;
    @SerializedName("niceDate")
    private String niceDate;
    @SerializedName("niceShareDate")
    private String niceShareDate;
    @SerializedName("envelopePic")
    private String envelopePic;
    @SerializedName("chapterName")
    private String chapterName;
    @SerializedName("superChapterName")
    private String superChapterName;
    @SerializedName("zan")
    private int zan;
    @SerializedName("collect")
    private boolean collect;

    public Article() {
    }

    public Article(int id, String title, String summary, String source,
                   String sourceUrl, String publishTime, String imageUrl,
                   String category, int commentCount, boolean favorite) {
        this.id = id;
        this.title = title;
        this.summary = summary;
        this.source = source;
        this.sourceUrl = sourceUrl;
        this.publishTime = publishTime;
        this.imageUrl = imageUrl;
        this.category = category;
        this.commentCount = commentCount;
        this.favorite = favorite;
    }

    public Article normalize() {
        //去除Html标签
        title = HtmlUtils.plainText(title);
        if (isBlank(title)) {
            title = "暂无标题";
        }

        if (isBlank(summary)) {
            String normalizedDesc = HtmlUtils.plainText(apiDesc);
            if (!isBlank(normalizedDesc)) {
                summary = normalizedDesc;
            } else {
                String normalizedCategory = buildCategory();
                summary = isBlank(normalizedCategory) ? "暂无摘要" : "分类：" + normalizedCategory;
            }
        } else {
            summary = HtmlUtils.plainText(summary);
        }

        if (isBlank(source)) {
            source = !isBlank(author) ? author : (!isBlank(shareUser) ? shareUser : "WanAndroid");
        }

        if (isBlank(sourceUrl)) {
            sourceUrl = link;
        }

        if (isBlank(publishTime)) {
            publishTime = !isBlank(niceDate) ? niceDate : niceShareDate;
        }

        if (isBlank(imageUrl)) {
            imageUrl = envelopePic;
        }

        if (isBlank(category)) {
            category = buildCategory();
        }

        if (commentCount <= 0) {
            commentCount = Math.max(zan, 0);
        }

        favorite = favorite || collect;
        return this;
    }

    private String buildCategory() {
        if (!isBlank(superChapterName) && !isBlank(chapterName)) {
            return superChapterName + " / " + chapterName;
        }
        if (!isBlank(superChapterName)) {
            return superChapterName;
        }
        return chapterName;
    }

    private boolean isBlank(String value) {
        return value == null || value.trim().isEmpty();
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getTitle() {
        return title == null ? "" : title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getSummary() {
        return summary == null ? "" : summary;
    }

    public void setSummary(String summary) {
        this.summary = summary;
    }

    public String getSource() {
        return source == null ? "" : source;
    }

    public void setSource(String source) {
        this.source = source;
    }

    public String getSourceUrl() {
        return sourceUrl == null ? "" : sourceUrl;
    }

    public void setSourceUrl(String sourceUrl) {
        this.sourceUrl = sourceUrl;
    }

    public String getPublishTime() {
        return publishTime == null ? "" : publishTime;
    }

    public void setPublishTime(String publishTime) {
        this.publishTime = publishTime;
    }

    public String getImageUrl() {
        return imageUrl == null ? "" : imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public String getCategory() {
        return category == null ? "" : category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public int getCommentCount() {
        return commentCount;
    }

    public void setCommentCount(int commentCount) {
        this.commentCount = commentCount;
    }

    public boolean isFavorite() {
        return favorite;
    }

    public void setFavorite(boolean favorite) {
        this.favorite = favorite;
    }
}

File: app/src/main/java/com/compuspulse/data/remote/model/Article.kt
package com.compuspulse.data.remote.model

import com.compuspulse.utils.HtmlUtils
import com.google.gson.annotations.SerializedName
import java.io.Serializable
import kotlin.math.max

class Article() : Serializable {
    private var id: Int = 0
    private var title: String? = null
    private var summary: String? = null
    private var source: String? = null
    private var sourceUrl: String? = null
    private var publishTime: String? = null
    private var imageUrl: String? = null
    private var category: String? = null
    private var commentCount: Int = 0
    private var favorite: Boolean = false

    @SerializedName("originId")
    private var originId: Int = 0

    @SerializedName("desc")
    private var apiDesc: String? = null

    @SerializedName("author")
    private var author: String? = null

    @SerializedName("shareUser")
    private var shareUser: String? = null

    @SerializedName("link")
    private var link: String? = null

    @SerializedName("niceDate")
    private var niceDate: String? = null

    @SerializedName("niceShareDate")
    private var niceShareDate: String? = null

    @SerializedName("envelopePic")
    private var envelopePic: String? = null

    @SerializedName("chapterName")
    private var chapterName: String? = null

    @SerializedName("superChapterName")
    private var superChapterName: String? = null

    @SerializedName("zan")
    private var zan: Int = 0

    @SerializedName("collect")
    private var collect: Boolean = false

    constructor(
        id: Int,
        title: String?,
        summary: String?,
        source: String?,
        sourceUrl: String?,
        publishTime: String?,
        imageUrl: String?,
        category: String?,
        commentCount: Int,
        favorite: Boolean
    ) : this() {
        this.id = id
        this.title = title
        this.summary = summary
        this.source = source
        this.sourceUrl = sourceUrl
        this.publishTime = publishTime
        this.imageUrl = imageUrl
        this.category = category
        this.commentCount = commentCount
        this.favorite = favorite
    }

    fun normalize(): Article {
        //去除Html标签
        title = HtmlUtils.plainText(title)
        if (isBlank(title)) {
            title = "暂无标题"
        }

        if (isBlank(summary)) {
            val normalizedDesc = HtmlUtils.plainText(apiDesc)
            summary = if (!isBlank(normalizedDesc)) {
                normalizedDesc
            } else {
                val normalizedCategory = buildCategory()
                if (isBlank(normalizedCategory)) "暂无摘要" else "分类：$normalizedCategory"
            }
        } else {
            summary = HtmlUtils.plainText(summary)
        }

        if (isBlank(source)) {
            source = if (!isBlank(author)) author else if (!isBlank(shareUser)) shareUser else "WanAndroid"
        }

        if (isBlank(sourceUrl)) {
            sourceUrl = link
        }

        if (isBlank(publishTime)) {
            publishTime = if (!isBlank(niceDate)) niceDate else niceShareDate
        }

        if (isBlank(imageUrl)) {
            imageUrl = envelopePic
        }

        if (isBlank(category)) {
            category = buildCategory()
        }

        if (commentCount <= 0) {
            commentCount = max(zan, 0)
        }

        favorite = favorite || collect
        return this
    }

    private fun buildCategory(): String? {
        if (!isBlank(superChapterName) && !isBlank(chapterName)) {
            return "$superChapterName / $chapterName"
        }
        if (!isBlank(superChapterName)) {
            return superChapterName
        }
        return chapterName
    }

    private fun isBlank(value: String?): Boolean = value == null || value.trim().isEmpty()

    fun getId(): Int = id

    fun setId(id: Int) {
        this.id = id
    }

    fun getTitle(): String = title ?: ""

    fun setTitle(title: String?) {
        this.title = title
    }

    fun getSummary(): String = summary ?: ""

    fun setSummary(summary: String?) {
        this.summary = summary
    }

    fun getSource(): String = source ?: ""

    fun setSource(source: String?) {
        this.source = source
    }

    fun getSourceUrl(): String = sourceUrl ?: ""

    fun setSourceUrl(sourceUrl: String?) {
        this.sourceUrl = sourceUrl
    }

    fun getPublishTime(): String = publishTime ?: ""

    fun setPublishTime(publishTime: String?) {
        this.publishTime = publishTime
    }

    fun getImageUrl(): String = imageUrl ?: ""

    fun setImageUrl(imageUrl: String?) {
        this.imageUrl = imageUrl
    }

    fun getCategory(): String = category ?: ""

    fun setCategory(category: String?) {
        this.category = category
    }

    fun getCommentCount(): Int = commentCount

    fun setCommentCount(commentCount: Int) {
        this.commentCount = commentCount
    }

    fun isFavorite(): Boolean = favorite

    fun setFavorite(favorite: Boolean) {
        this.favorite = favorite
    }

    fun getOriginId(): Int = originId

    fun getServerArticleId(): Int = if (originId > 0) originId else id
}
